### PR TITLE
PSD-1001: Create file upload component

### PR DESCRIPTION
**Pull request by kicferk1**
Adds file upload component mirroring govuk macros, as well as Rails glue and Document model that we use to make the glue work. 

This includes a re-work of file validation, which now lives exclusively on the Document model. Likewise, updating, saving and attaching files lives there, because it seemed like a better place than a controller concern. 

It also adds a way of dealing with displaying error summary for a form that has one or more subforms. Unfortunately, I'm not sure there is a way of making it work with custom ids for validated fields(and I have a sneaking suspicion current error_summary component isn't well equipped to fix that). Which means we might have to rework `related_attachment_fields` to get the desired 'on click page scrolls to error message' behaviour. 

Still to do: Add a gallery, figure our error linking when field gets custom id, or remove custom ids. Probably move the document model to shared-web, because without it the glue component doesn't really work. 

## Checklist:
- [x] Automated checks are passing locally.
### General testing
- [x] Test without javascript
- [x] Test on small screen
### Accessibility testing
- [x] Works keyboard only
- [x] Tested with one screen reader
- [x] Zoom page to 400% - content still visible
- [x] Disable css - does content make sense and appear in a logical order?
- [x] Passes automated checker (automated in build or manual)

### CODE REVIEW

File: psd-web/app/views/application/_related_attachment_fields.html.slim [hunk 1]
```diff
@@ -2,15 +2,14 @@ fieldset.govuk-fieldset
...
+      ' Currently selected file:
+      = link_to file_model.filename, file_model.get_blob, target: "_blank", rel: 'noopener'
+  - else
+    = form.fields_for attachment_name, file_model do |subform|
+      = render "form_components/govuk_file_upload", form: subform, key: :file, id: "attachment-file-input",
+              label: { text: "File upload", class: "govuk-visually-hidden" }
       #attachment-description
-        = render "form_components/govuk_textarea", key: :description, value: file_blob&.metadata&.dig(:description),
-              form: subform, label: { text: "Attachment description" }, attributes: { maxlength: 10000 }
+        = render "form_components/govuk_textarea", key: :description, form: subform,
```
Comment (jasiekmiko) [inline psd-web/app/views/application/_related_attachment_fields.html.slim]: Does the error-summary used on these pages capture errors on this subform?

File: psd-web/app/views/documents/remove.html.slim [hunk 2]
```diff
@@ -12,22 +12,22 @@
           th.govuk-table__header[scope="row"]
             | Title
           td.govuk-table__cell
-            = @file.metadata[:title]
+            = @file_model.title
```
Comment (jasiekmiko) [inline psd-web/app/views/documents/remove.html.slim]: so much nicer 😄

File: psd-web/app/views/investigations/corrective_actions/confirmation.html.slim [hunk 3]
```diff
@@ -29,14 +29,14 @@ h1.govuk-heading-l
           tr.govuk-table__row
             th.govuk-table__header scope="row" Details
             td.govuk-table__cell = @corrective_action.details
-        - if @file_blob
+        - if @file_model.get_blob
```
Comment (jasiekmiko) [inline psd-web/app/views/investigations/corrective_actions/confirmation.html.slim]: Should all of these variables be renamed to "document" to match the model name? Or should the model be called "file"?

File: psd-web/app/views/investigations/ts_investigations/_file_page.html.slim [hunk 4]
```diff
@@ -3,18 +3,18 @@ fieldset.govuk-fieldset
...
-          = subform.file_field :file, class: "govuk-file-upload", id: "attachment-file-input"
-      hr.govuk-section-break.govuk-section-break--m.govuk-section-break--visible
-      = render "form_components/govuk_input", key: :title, form: subform, value: @file_title,
-              classes: "govuk-!-width-two-thirds",
-              label: { text: "Title" }
-      = render "form_components/govuk_textarea", key: :description, form: subform, value: @file_description,
-              attributes: { maxlength: 10000 }, label: { text: "Description" },
-              hint: { text: "For example, 'Orange, 4 socket extension lead'" }
+        = link_to @file_model.filename, @file_model.get_blob, target: "_blank", rel: 'noopener'
+    - else
```
Comment (jasiekmiko) [inline psd-web/app/views/investigations/ts_investigations/_file_page.html.slim]: Why is that an `else`? I thought we always show the input, but annotate it with current file link when it's there?

File: shared-web/app/views/form_components/_govuk_error_summary.html.slim [hunk 5]
```diff
@@ -7,5 +7,17 @@ ruby:
     prefix = get_attribute_id_prefix(form, attribute)
     errors.full_messages_for(attribute).map { |m| { text: m, href: "##{prefix}-error" } }
   end
-- if errors.any?
+
+  if local_assigns[:subform_data].present?
```
Comment (jasiekmiko) [inline shared-web/app/views/form_components/_govuk_error_summary.html.slim]: This doesn't feel like a very smooth way to use the glue - it requires very explicit knowledge of the object when calling the `error_summary` component. How would rails achieve this? Can we copy that approach?

File: shared-web/app/views/form_components/_govuk_file_upload.html.slim [hunk 6]
```diff
@@ -0,0 +1,15 @@
+/ This partial acts as glue between rails form helpers and the component file upload
+/ With access to the form object, we can infer a lot of information from it
+
+ruby:
+  # Infer name and id_prefix from attribute key
+  form.multipart = true
```
Comment (jasiekmiko) [inline shared-web/app/views/form_components/_govuk_file_upload.html.slim]: Explaining the reason for this would be neat - especially with a reference to where in rails this is normally done

File: psd-web/app/assets/application/javascripts/investigations/attachment_description.js [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import $ from 'jquery';
 
 $(document).ready(() => {
-  const attachmentFileInput = document.getElementById('attachment-file-input');
+  const attachmentFileInput = $(document.getElementById('attachment-file-input')).find('input')[0];
```
Comment (jasiekmiko) [inline psd-web/app/assets/application/javascripts/investigations/attachment_description.js]: If using jquery, then let's use it's full power:
`$('#attachment-file-input input')[0]` or so

File: psd-web/app/models/document.rb [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  attr_accessor :filename
+  attr_accessor :required_fields
+  attr_accessor :attachment
+
+  validate :validate_blob_size
+  validate :has_required_fields
+
+  def initialize(attachment, required_fields = [])
+    @file = attachment.is_a?(ActiveStorage::Attachment) ? attachment.blob_id : attachment.id if attachment
+    @attachment = attachment.is_a?(ActiveStorage::Attachment) ? attachment : nil
```
[the remark this answers is not in the document]
Comment (kicferk1) [reply psd-web/app/models/document.rb]: I added this to avoid a bug with id on edit/remove file pages. But it can be fixed in a better way.

File: psd-web/app/models/document.rb [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    @title = attachment.metadata["title"] if attachment&.metadata
```
[the remark this answers is not in the document]
Comment (kicferk1) [reply psd-web/app/models/document.rb]: A drawback of starting to change everything with `document_controller` is accommodating or its weirdness. 

It's the only place that would like to pass an attachment to document, so it's easier to enforce receiving a blob rather than meaningfully allow constructor for either. 

The meaning for no-attachment document is actually pretty simple, it's any document that has been uploaded(in a flow), but not attached yet(because the flow wasn't completed).

File: psd-web/app/models/document.rb [hunk 10]
```diff
@@ -0,0 +1,100 @@
+class Document
```
[the remark this answers is not in the document]
Comment (kicferk1) [reply psd-web/app/models/document.rb]: Where should it live? Comments before methods it affects?

File: psd-web/app/models/document.rb [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  end
+
+  def validate_blob_size
+    blob = get_blob
+    return unless blob && (blob.byte_size > max_file_byte_size)
+
+    errors.add(:base, :file_too_large, message: "File is too big, allowed size is #{max_file_byte_size / 1.megabyte} MB")
+  end
+
+  def max_file_byte_size
```
[the remark this answers is not in the document]
Comment (kicferk1) [reply psd-web/app/models/document.rb]: I didn't plan for it to be overridden, just moved it from a controller concern without giving it much thought. 

It used to be a method specifically so that a controller could override it if it wanted to. Then again, no controller ever did, and it has been a long while since it was implemented. 

It could be overridden, say if we create `BigDocument < Document`, and a controller uses that new model to construct its `@file_model`. I doubt we will use it in a foreseeable future, but cosmetics might.

File: psd-web/app/views/application/_related_attachment_fields.html.slim (hunk 1 above)
Comment (kicferk1) [reply psd-web/app/views/application/_related_attachment_fields.html.slim]: Yes

File: psd-web/app/models/document.rb [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+  attr_accessor :description
+  attr_accessor :document_type
+  attr_accessor :filename
+  attr_accessor :required_fields
+  attr_accessor :attachment
+
+  validate :validate_blob_size
+  validate :has_required_fields
+
+  def initialize(attachment, required_fields = [])
```
[the remark this answers is not in the document]
Comment (kicferk1) [reply psd-web/app/models/document.rb]: I'll add tests anyway, but I think document should actually only accept an active storage blob(or nil).

[the remark this answers is not in the document]
Comment (jasiekmiko) [reply psd-web/app/models/document.rb]: That sounds like a good plan. What's the `nil` usecase?

File: psd-web/app/models/test/result.rb [hunk 13]
```diff
@@ -2,7 +2,6 @@ class Test::Result < Test
   after_create :create_audit_activity
 
   validates :result, presence: { message: "Select result of the test" }
-  validate :result_file_attached
```
[the remark this answers is not in the document]
Comment (kicferk1) [reply psd-web/app/models/test/result.rb]: The way the behaviour is preserved is by controller setting correct required fields in document model that it then validates. 

As for using magic like `has_many :documents`, and building validation on top of that, I'm not sure it's worth the trouble :/

File: psd-web/app/views/investigations/ts_investigations/_file_page.html.slim (hunk 4 above)
Comment (jasiekmiko) [reply psd-web/app/views/investigations/ts_investigations/_file_page.html.slim]: oh god. i suspect we should, o/w how do you change if you uploaded the wrong file?
Probably not one for this PR though..

File: psd-web/app/models/document.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (jasiekmiko) [reply psd-web/app/models/document.rb]: Cool, sounds like we want to simplify it down to a class constant then and worry about making it variable when we need to

File: psd-web/app/models/document.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (kicferk1) [reply psd-web/app/models/document.rb]: One use case is to validate that a file was uploaded, and if it wasn't, to attach an error to this model in `:file` field, so that error handling can work its magic smoothly. 

The other is optional file inputs, if a file doesn't need to be attached, this model should still work to remove some complexity from controllers.

File: psd-web/app/controllers/concerns/corrective_actions_concern.rb [hunk 14]
```diff
@@ -16,20 +16,19 @@ def corrective_action_params
   end
 
   def set_attachment
-    @file_blob, * = load_file_attachments
-    if @file_blob && @corrective_action.related_file == "Yes"
-      @corrective_action.documents.attach(@file_blob)
+    file_blob, * = load_file_attachments
+    @file_model = Document.new(file_blob)
```
[the remark this answers is not in the document]
Comment (kicferk1) [reply psd-web/app/controllers/concerns/corrective_actions_concern.rb]: I don't want the controller to have to worry about `@file_model ` being nil :/

File: psd-web/app/models/document.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (jasiekmiko) [reply psd-web/app/models/document.rb]: It might be worth making that explicit in the comments :)

File: shared-web/app/views/form_components/_govuk_error_summary.html.slim (hunk 5 above)
Comment (kicferk1) [reply shared-web/app/views/form_components/_govuk_error_summary.html.slim]: As magical as rails is, I don't think it has a nice way of getting around that. 

At the time `error_summary` is generated, the main form doesn't know there will be a subform defined further down the line. So we can't recursively compute errors for all subforms. 

After prodding rails ObjectSpace, it seems that at the time we call `error_summary`(the top of each main form), not even rails knows there is a subform object. Which probably makes sense, initialisation happens when `fields_for` is called, which is after `error_summary` gets generated.

A solution would be to create subform before error summary, but I think that defeats the purpose of 'needing to explicitly know the object'. 

Another solution would be to try and attach copies of subform object errors to main form object, and naming their keys cleverly so that href still works. But that would require adding these attributes to parent models just to accommodate errors :(

File: psd-web/app/models/test/result.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (jasiekmiko) [reply psd-web/app/models/test/result.rb]: Hmm, I think you're right about the `has_many`, at least for the time being.
But is controller definitely the right place to control the behaviour of the model? I think the property of what makes a valid model should live within the model(s) and the controllers should just be able to tap into that knowledge.
Is it worth having a `validate :attached_documents` which triggers the validation of the attached documents? It could be what sets the required fields too, perhaps?

File: psd-web/app/models/document.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (jasiekmiko) [reply psd-web/app/models/document.rb]: yeah, sounds like a good place - epsecially if it's referencing methods in the class or whatnot

File: shared-web/app/views/form_components/_govuk_error_summary.html.slim (hunk 5 above)
Comment (jasiekmiko) [reply shared-web/app/views/form_components/_govuk_error_summary.html.slim]: Do we need the subform? Or do we just need to know that there are children objects on this main one?

Comment (jasiekmiko) [reply shared-web/app/views/form_components/_govuk_error_summary.html.slim]: is a variation on this going to be helpful? https://stackoverflow.com/a/2718330/2363767

File: psd-web/app/controllers/concerns/corrective_actions_concern.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (kicferk1) [reply psd-web/app/controllers/concerns/corrective_actions_concern.rb]: But then when I call `@file_model.validate`, or `@file_model.attach_blob_to_list` I need to add `&`s to make sure nothing breaks. And our views need to worry about it being nil as well :/

In short, it's a lot easier to push worrying about file blob being nil to document model, and it simplifies this bit of logic as well(check out new commits :) )

[the remark this answers is not in the document]
Comment (jasiekmiko) [reply psd-web/app/controllers/concerns/corrective_actions_concern.rb]: Yeah no you're completely right, I was being silly and ignoring the `@` in that variable 🤦‍♂️

File: psd-web/app/controllers/documents_controller.rb [hunk 15]
```diff
@@ -15,43 +15,32 @@ def edit; end
...
-  def file_valid?
-    if @file.blank? || @file.blob.blank?
-      @errors.add(:base, :file_not_implemented, message: "File can't be blank")
-    end
-    if @file.metadata[:title].blank?
-      @errors.add(:base, :title_not_implemented, message: "Title can't be blank")
-    end
-    validate_blob_size(@file, @errors, "file")
-    @errors.empty?
+    @file_model = Document.new(file, [[:title, "Enter title"]])
```
[the remark this answers is not in the document]
Comment (kicferk1) [reply psd-web/app/controllers/documents_controller.rb]: We don't need customisation(I think), I used `:required` instead of `:blank` as type of error, which resulted in ugly messages :P

**Comment by kicferk1** [comment]
Since I'm leaving the project, and last commit includes a lot of magic, here is some explanation of what it does. 

Controller shouldn't care about validation. So it shouldn't set required fields of a model. A model should set those itself. So I created a concern that should be included in models that can have things uploaded to them. It includes a macro `validate_upload_document key, required, context=nil`, which does just that. 

`key` should be the key of subform we upload document in. `required` should be an array of required fields in that subform. `context` should be, well, context, so a wizard step, etc. ~

It also adds attribute `key` to the model. It will use it to store the document model. 

Controller needs to set the document model on an object. 

Example of usage is in the last commit, in `enquiry.rb` and `creation_flow_controller.rb`.

**Comment by slorek** [comment]
Closing as not updated for several months.
